src/haxeLanguageServer/features/ExtractFunctionFeature.hx

```diff
@@ -0,0 +1,274 @@
+			switch (parent.tok) {
+				case Kwd(KwdFunction):
+					return parent;
+				default:
+			}
+			parent = parent.parent;
+		}
+		return null;
+	}
+
+	function makeReturnSpec(functionToken:TokenTree):String {
+		var returnHint:Null<TokenTree> = functionToken.access().firstChild().isCIdent().firstOf(DblDot).token;
+		// anon function
+		if (returnHint == null)
+			returnHint = functionToken.access().firstOf(DblDot).token;
+		if ((returnHint == null) || (returnHint.children == null))
+			return "";
+		return varToString(returnHint);
+	}
+
+	function copyParentFunctionParameters(functionToken:TokenTree, text:String, rangeIdents:Array<String>):Array<NewFunctionParameter> {
+		var paramterList:Null<TokenTree> = functionToken.access().firstChild().isCIdent().firstOf(POpen).token;
+		// anon function
+		if (paramterList == null)
+			paramterList = functionToken.access().firstOf(POpen).token;
+
+		if ((paramterList == null) || (paramterList.children == null))
+			return [];
+
+		var newFuncParameter:Array<NewFunctionParameter> = [];
+		for (child in paramterList.children) {
+			switch (child.tok) {
+				case Const(CIdent(s)):
+					checkAndAddIdentifier(child, s, text, rangeIdents, newFuncParameter);
+				case Question:
+					var firstChild:Null<TokenTree> = child.getFirstChild();
+					if (firstChild == null)
+						continue;
+					switch (firstChild.tok) {
+						case Const(CString(s)):
+							checkAndAddIdentifier(child, s, text, rangeIdents, newFuncParameter);
+						default:
+					}
+				case Dollar(s):
+					if (rangeIdents.indexOf("$" + s) < 0)
+						continue;
+					newFuncParameter.push({
+						call: s,
+						param: varToString(child)
+					});
+				case PClose:
+					return newFuncParameter;
+				default:
+			}
+		}
+
+		return newFuncParameter;
+	}
+
+	function checkAndAddIdentifier(token:TokenTree, identifier:String, text:String, rangeIdents:Array<String>, newFuncParameter:Array<NewFunctionParameter>) {
+		if (rangeIdents.indexOf(identifier) >= 0)
+			newFuncParameter.push({
+				call: identifier,
+				param: varToString(token)
+			});
+		if (text.indexOf("$" + identifier) >= 0)
+			newFuncParameter.push({
+				call: identifier,
+				param: varToString(token)
+			});
+	}
+
+	function localVarsToParameter(varTokens:Array<TokenTree>, text:String, rangeIdents:Array<String>):Array<NewFunctionParameter> {
+		var newFuncParameter:Array<NewFunctionParameter> = [];
+
+		for (varToken in varTokens) {
+			// TODO handle multiple vars
+			for (child in varToken.children) {
+				switch (child.tok) {
+					case Const(CIdent(s)):
+						checkAndAddIdentifier(child, s, text, rangeIdents, newFuncParameter);
+					case Dollar(s):
+						if (rangeIdents.indexOf("$" + s) < 0)
+							continue;
+						newFuncParameter.push({
+							call: s,
+							param: varToString(child)
+						});
+					default:
+						continue;
+				}
+			}
+		}
+		return newFuncParameter;
+	}
+
+	function varToString(token:TokenTree):String {
```

Comment:
Hm... Seems like this probably doesn't belong here long-term... we might want to have some `TokenTreePrinter` utility class in the token-tree lib?

Reply:
well, here we want `x:Type<T>` but no assignment, no comma, no semicolon... a generic TokenTreePrinter would produce the full subtree, which can't be placed inside a function signature